src/main/java/com/selfxdsd/todos/PuzzlesComponent.java

```diff
@@ -57,153 +50,68 @@
-                author = "@" + author + " ";
-            } else {
-                author = "";
-            }
-            commit.comments().post(
-                author + "I've opened the Issues "
-                + opened + " for the newly added to-dos.\n\n"
-                + "The to-dos may have been added in an earlier commit, "
-                + "but I've found them just now."
-            );
-        }
+        this.beanFactory = beanFactory;
+        this.reviewer = reviewer;
     }
 
     /**
-     * Close issues which don't have a corresponding puzzle
-     * (puzzle has been removed from code).
-     * @param puzzles Puzzles found in the repo.
-     * @param issues Issues API.
-     * @param commit Commit which triggered everything.
+     * Listener for {@link PuzzlesWebhookEvent} Spring event triggered
+     * on webhook request.
+     * @param event Puzzles Webhook Event.
+     **@throws Exception If something goes wrong on closing database connection.
      */
-    private void closeRemovedPuzzles(
-        final Puzzles<Project> puzzles,
-        final Issues issues,
-        final Commit commit
-    ) {
-        final List<String> closed = new ArrayList<>();
-        for(final Issue issue : issues) {
-            boolean foundPuzzle = false;
-            for(final Puzzle puzzle : puzzles) {
-                if(issue.body().contains(puzzle.getId())) {
-                    foundPuzzle = true;
-                    break;
-                }
-            }
-            if(!foundPuzzle && !issue.isClosed()) {
-                issue.close();
-                issue.comments().post(
-                    "Puzzle disappeared from the code, "
-                    + "that's why I closed this ticket."
-                );
-                closed.add("#" + issue.issueId());
-            }
-        }
-        if(closed.size() > 0) {
-            String author = commit.author();
-            if(author != null && !author.isEmpty()) {
-                author = "@" + author + " ";
+    @Async
+    @EventListener
+    public void onPuzzleWebhookEvent(final PuzzlesWebhookEvent event)
+        throws Exception {
+        final Self selfCore = this.beanFactory.getBean(Self.class);
+        final String provider = event.getProvider();
+        final String repoFullName = event.getProjectFullName();
+        final Project project = selfCore.projects().getProjectById(
+            repoFullName, provider
+        );
+        if (project != null) {
+            if (provider.equalsIgnoreCase(Provider.Names.GITHUB)) {
+                this.reviewer.review(WebhookEvents.create(project,
+                    "push",
+                    event.getPayload()
+                ));
+            } else if (provider.equalsIgnoreCase(Provider.Names.GITLAB)) {
+                this.reviewer.review(WebhookEvents.create(
+                    project,
+                    "Push Hook",
+                    event.getPayload()
+                ));
```

Comment:
now is safe to create the 'heavy' Self Event and passing it to a reviewer (PuzzlesReviewer implementation) that's going to do all the heavylifting that PuzzlesComponent did before.